components/layout/display_list_builder.rs

```diff
@@ -1184,172 +1362,103 @@ impl FragmentDisplayListBuilding for Fragment {
         }
     }
 
-    fn compute_background_image_size(&self,
-                                     style: &ComputedValues,
-                                     bounds: &Rect<Au>,
-                                     image: &WebRenderImageInfo,
-                                     index: usize)
-                                     -> Size2D<Au> {
-        // If `image_aspect_ratio` < `bounds_aspect_ratio`, the image is tall; otherwise, it is
-        // wide.
-        let image_aspect_ratio = (image.width as f64) / (image.height as f64);
-        let bounds_aspect_ratio = bounds.size.width.to_f64_px() / bounds.size.height.to_f64_px();
-        let intrinsic_size = Size2D::new(Au::from_px(image.width as i32),
-                                         Au::from_px(image.height as i32));
-        let background_size = get_cyclic(&style.get_background().background_size.0, index).clone();
-        match (background_size, image_aspect_ratio < bounds_aspect_ratio) {
-            (BackgroundSize::Contain, false) | (BackgroundSize::Cover, true) => {
-                Size2D::new(bounds.size.width,
-                            Au::from_f64_px(bounds.size.width.to_f64_px() / image_aspect_ratio))
-            }
-
-            (BackgroundSize::Contain, true) | (BackgroundSize::Cover, false) => {
-                Size2D::new(Au::from_f64_px(bounds.size.height.to_f64_px() * image_aspect_ratio),
-                            bounds.size.height)
-            }
+    fn compute_background_placement(&self,
+                                    state: &mut DisplayListBuildState,
+                                    style: &ComputedValues,
+                                    absolute_bounds: Rect<Au>,
+                                    intrinsic_size: Option<Size2D<Au>>,
+                                    index: usize)
+                                    -> (Rect<Au>,
```

Comment:
nit: Maybe a struct would be nicer instead of a tuple? It's not clear what each field means without looking at the function implementation.